File: rust/cx-sdk/src/client/archive_retentions.rs
// Copyright 2024 Coralogix Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use cx_api::proto::com::coralogix::archive::v1::{
    ActivateRetentionsRequest,
    ActivateRetentionsResponse,
    GetRetentionsEnabledRequest,
    GetRetentionsEnabledResponse,
    GetRetentionsRequest,
    GetRetentionsResponse,
    UpdateRetentionsRequest,
    UpdateRetentionsResponse,
    retentions_service_client::RetentionsServiceClient,
};
use std::str::FromStr;
use tokio::sync::Mutex;
use tonic::{
    metadata::MetadataMap,
    transport::{
        Channel,
        ClientTlsConfig,
        Endpoint,
    },
};

use crate::{
    CoralogixRegion,
    auth::AuthContext,
    error::{
        Result,
        SdkApiError,
        SdkError,
    },
    metadata::CallProperties,
    util::make_request_with_metadata,
};

pub use cx_api::proto::com::coralogix::archive::v1::RetentionUpdateElement;

const ARCHIVE_FEATURE_GROUP_ID: &str = "archive";

/// The Archive Retention API client.
/// Read more at <https://coralogix.com/docs/archive-retention/>
pub struct ArchiveRetentionClient {
    metadata_map: MetadataMap,
    service_client: Mutex<RetentionsServiceClient<Channel>>,
}

impl ArchiveRetentionClient {
    /// Creates a new client for the Archive Retention API.
    ///
    /// # Arguments
    /// * `auth_context` - The [`AuthContext`] to use for authentication.
    /// * `region` - The [`CoralogixRegion`] to connect to.
    pub fn new(auth_context: AuthContext, region: CoralogixRegion) -> Result<Self> {
        let channel: Channel = Endpoint::from_str(region.grpc_endpoint().as_str())?
            .tls_config(ClientTlsConfig::new().with_native_roots())?
            .connect_lazy();
        let request_metadata: CallProperties = (&auth_context.team_level_api_key).into();
        Ok(Self {
            metadata_map: request_metadata.to_metadata_map(),
            service_client: Mutex::new(RetentionsServiceClient::new(channel)),
        })
    }

    /// Retrieves the retention settings for a tenant.
    pub async fn get(&self) -> Result<GetRetentionsResponse> {
        Ok(self
            .service_client
            .lock()
            .await
            .get_retentions(make_request_with_metadata(
                GetRetentionsRequest {},
                &self.metadata_map,
            ))
            .await
            .map_err(|status| {
                SdkError::ApiError(SdkApiError {
                    status,
                    endpoint: "/com.coralogix.archive.v1.RetentionsService/GetRetentions"
                        .to_string(),
                    feature_group: ARCHIVE_FEATURE_GROUP_ID.to_string(),
                })
            })?
            .into_inner())
    }

    /// Retrieves the enabled retention settings for a tenant.
    pub async fn get_enabled(&self) -> Result<GetRetentionsEnabledResponse> {
        Ok(self
            .service_client
            .lock()
            .await
            .get_retentions_enabled(make_request_with_metadata(
                GetRetentionsEnabledRequest {},
                &self.metadata_map,
            ))
            .await
            .map_err(|status| {
                SdkError::ApiError(SdkApiError {
                    status,
                    endpoint: "/com.coralogix.archive.v1.RetentionsService/GetEnabled".to_string(),
                    feature_group: ARCHIVE_FEATURE_GROUP_ID.to_string(),
                })
            })?
            .into_inner())
    }

    /// Updates the retention settings for a tenant.
    pub async fn update(
        &self,
        retention_update_elements: Vec<RetentionUpdateElement>,
    ) -> Result<UpdateRetentionsResponse> {
        let response = {
            let mut client = self.service_client.lock().await;
            let update_response = client
                .update_retentions(make_request_with_metadata(
                    UpdateRetentionsRequest {
                        retention_update_elements,
                    },
                    &self.metadata_map,
                ))
                .await;
            match update_response.map_err(|status| {
                SdkError::ApiError(SdkApiError {
                    status,
                    endpoint: "/com.coralogix.archive.v1.RetentionsService/UpdateRetentions"
                        .to_string(),
                    feature_group: ARCHIVE_FEATURE_GROUP_ID.to_string(),
                })
            }) {
                Ok(output) => {
                    let _ = client
                        .activate_retentions(ActivateRetentionsRequest {})
                        .await
                        .map_err(|status| {
                            SdkError::ApiError(SdkApiError {
                                status,
                                endpoint:
                                    "/com.coralogix.archive.v1.RetentionsService/ActivateRetentions"
                                        .to_string(),
                                feature_group: ARCHIVE_FEATURE_GROUP_ID.to_string(),
                            })
                        })?;
                    output.into_inner()
                }
                Err(error) => return Err(error),
            }
        };
        Ok(response)
    }

    /// Activates all archive retentions    
    #[deprecated(since = "1.8.0", note = "Update activates automatically")]
    pub async fn activate(&self) -> Result<ActivateRetentionsResponse> {
        Ok(self
            .service_client
            .lock()
            .await
            .activate_retentions(make_request_with_metadata(
                ActivateRetentionsRequest {},
                &self.metadata_map,
            ))
            .await
            .map_err(|status| {
                SdkError::ApiError(SdkApiError {
                    status,
                    endpoint: "/com.coralogix.archive.v1.RetentionsService/ActivateRetentions"
                        .to_string(),
                    feature_group: ARCHIVE_FEATURE_GROUP_ID.to_string(),
                })
            })?
            .into_inner())
    }
}

File: go/archive-retentions.go
// Copyright 2024 Coralogix Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package cxsdk

import (
	"context"

	archiveRetention "github.com/coralogix/coralogix-management-sdk/go/internal/coralogix/archive/v1"
)

// UpdateRetentionsRequest is a type for archive retention.
type UpdateRetentionsRequest = archiveRetention.UpdateRetentionsRequest

// GetRetentionsRequest is a type for archive retention.
type GetRetentionsRequest = archiveRetention.GetRetentionsRequest

// RetentionUpdateElement is a type for archive retention.
type RetentionUpdateElement = archiveRetention.RetentionUpdateElement

// GetRetentionsEnabledRequest is a request type.
type GetRetentionsEnabledRequest = archiveRetention.GetRetentionsEnabledRequest

// ActivateRetentionsRequest is a request type.
type ActivateRetentionsRequest = archiveRetention.ActivateRetentionsRequest

// Retention is a type for archive retention.
type Retention = archiveRetention.Retention

const archiveFeatureGroupID = "archive"

// RPC values.
const (
	ArchiveRetentionGetRetentionsRPC        = archiveRetention.RetentionsService_GetRetentions_FullMethodName
	ArchiveRetentionUpdateRetentionsRPC     = archiveRetention.RetentionsService_UpdateRetentions_FullMethodName
	ArchiveRetentionActivateRetentionsRPC   = archiveRetention.RetentionsService_ActivateRetentions_FullMethodName
	ArchiveRetentionGetRetentionsEnabledRPC = archiveRetention.RetentionsService_GetRetentionsEnabled_FullMethodName
)

// ArchiveRetentionsClient is a client for the Coralogix Archive Retentions API.
type ArchiveRetentionsClient struct {
	callPropertiesCreator *CallPropertiesCreator
}

// Get gets the archive retentions.
func (c ArchiveRetentionsClient) Get(ctx context.Context, req *archiveRetention.GetRetentionsRequest) (*archiveRetention.GetRetentionsResponse, error) {
	callProperties, err := c.callPropertiesCreator.GetTeamsLevelCallProperties(ctx)
	if err != nil {
		return nil, err
	}

	conn := callProperties.Connection
	defer conn.Close()
	client := archiveRetention.NewRetentionsServiceClient(conn)

	response, err := client.GetRetentions(callProperties.Ctx, req, callProperties.CallOptions...)
	if err != nil {
		return nil, NewSdkAPIError(err, ArchiveRetentionGetRetentionsRPC, archiveFeatureGroupID)
	}
	return response, nil
}

// Update updates the archive retentions.
func (c ArchiveRetentionsClient) Update(ctx context.Context, req *UpdateRetentionsRequest) (*archiveRetention.UpdateRetentionsResponse, error) {
	callProperties, err := c.callPropertiesCreator.GetTeamsLevelCallProperties(ctx)
	if err != nil {
		return nil, err
	}

	conn := callProperties.Connection
	defer conn.Close()
	client := archiveRetention.NewRetentionsServiceClient(conn)

	response, err := client.UpdateRetentions(callProperties.Ctx, req, callProperties.CallOptions...)
	if err != nil {
		return nil, NewSdkAPIError(err, ArchiveRetentionUpdateRetentionsRPC, archiveFeatureGroupID)
	}
	return response, nil
}

// Activate activates the archive retentions.
func (c ArchiveRetentionsClient) Activate(ctx context.Context, req *ActivateRetentionsRequest) (*archiveRetention.ActivateRetentionsResponse, error) {
	callProperties, err := c.callPropertiesCreator.GetTeamsLevelCallProperties(ctx)
	if err != nil {
		return nil, err
	}

	conn := callProperties.Connection
	defer conn.Close()
	client := archiveRetention.NewRetentionsServiceClient(conn)

	response, err := client.ActivateRetentions(callProperties.Ctx, req, callProperties.CallOptions...)
	if err != nil {
		return nil, NewSdkAPIError(err, ArchiveRetentionActivateRetentionsRPC, archiveFeatureGroupID)
	}
	return response, nil
}

// GetEnabled returns a boolean that signals whether archive retentions are enabled.
func (c ArchiveRetentionsClient) GetEnabled(ctx context.Context, req *GetRetentionsEnabledRequest) (*archiveRetention.GetRetentionsEnabledResponse, error) {
	callProperties, err := c.callPropertiesCreator.GetTeamsLevelCallProperties(ctx)
	if err != nil {
		return nil, err
	}

	conn := callProperties.Connection
	defer conn.Close()
	client := archiveRetention.NewRetentionsServiceClient(conn)

	response, err := client.GetRetentionsEnabled(callProperties.Ctx, req, callProperties.CallOptions...)
	if err != nil {
		return nil, NewSdkAPIError(err, ArchiveRetentionGetRetentionsEnabledRPC, archiveFeatureGroupID)
	}
	return response, nil

}

// NewArchiveRetentionsClient Creates a new archive retentions client.
func NewArchiveRetentionsClient(c *CallPropertiesCreator) *ArchiveRetentionsClient {
	return &ArchiveRetentionsClient{callPropertiesCreator: c}
}
